File: docs/source/_static/examples/cpp/orbitals.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE.txt in the project root for
// license information.

// Orbitals usage examples.
// --------------------------------------------------------------------------------------------
// docs:xyz ../data/h2.structure.xyz
// start-cell-create
#include <iostream>
#include <qdk/chemistry.hpp>
#include <string>
using namespace qdk::chemistry::data;
using namespace qdk::chemistry::algorithms;

int main() {
  // Obtain orbitals from a SCF calculation
  // Load H2 molecule from inline XYZ file
  auto structure = Structure::from_xyz(R"(2
H2 molecule
H    0.000000    0.000000    0.000000
H    0.000000    0.000000    0.740848
)");

  // Obtain orbitals from a SCF calculation
  auto scf_solver = ScfSolverFactory::create();
  auto [E_scf, wfn] = scf_solver->run(structure, 0, 1, "sto-3g");
  std::shared_ptr<Orbitals> orbitals = wfn.get_orbitals();

  // end-cell-create
  // --------------------------------------------------------------------------------------------

  // --------------------------------------------------------------------------------------------
  // start-cell-model-orbitals-create
  // Set basis set size
  size_t basis_size = 6;

  // Set active orbitals
  std::vector<size_t> alpha_active = {1, 2};
  std::vector<size_t> beta_active = {2, 3, 4};
  std::vector<size_t> alpha_inactive = {0, 3, 4, 5};
  std::vector<size_t> beta_inactive = {0, 1, 5};

  ModelOrbitals model_orbitals(
      basis_size, std::make_tuple(alpha_active, beta_active, alpha_inactive,
                                  beta_inactive));

  // We can then pass this object to a custom Hamiltonian constructor
  // end-cell-model-orbitals-create
  // --------------------------------------------------------------------------------------------

  // --------------------------------------------------------------------------------------------
  // start-cell-access
  // Access orbital coefficients as a symmetry-blocked tensor, then per spin
  // block
  auto coefficients = orbitals->coefficients();
  const auto& coeffs_alpha =
      coefficients->block({axes::alpha(), axes::alpha()});
  const auto& coeffs_beta = coefficients->block({axes::beta(), axes::beta()});

  // Access orbital energies as a symmetry-blocked tensor, then per spin block
  auto orbital_energies = orbitals->energies();
  const auto& energies_alpha = orbital_energies->block({axes::alpha()});
  const auto& energies_beta = orbital_energies->block({axes::beta()});

  // Get active space indices per spin channel
  auto active_indices_alpha =
      orbitals->active_indices()->indices(SymmetryLabel({axes::alpha()}));
  auto active_indices_beta =
      orbitals->active_indices()->indices(SymmetryLabel({axes::beta()}));

  // Access atomic orbital overlap matrix (returns const Eigen::MatrixXd&)
  const auto& ao_overlap = orbitals->get_overlap_matrix();

  // Access basis set information (returns const BasisSet&)
  const auto& basis_set = orbitals->get_basis_set();

  // Check calculation type
  bool is_restricted = orbitals->is_restricted();

  // Get size information
  size_t num_molecular_orbitals = orbitals->get_num_molecular_orbitals();
  size_t num_atomic_orbitals = orbitals->get_num_atomic_orbitals();

  std::string summary = orbitals->get_summary();
  std::cout << summary << std::endl;

  // end-cell-access
  // --------------------------------------------------------------------------------------------
  return 0;
}

File: docs/source/_static/examples/python/orbitals.py
"""Orbitals usage examples."""

# --------------------------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE.txt in the project root for license information.
# --------------------------------------------------------------------------------------------

################################################################################
# start-cell-create
import tempfile
from pathlib import Path

import numpy as np
from qdk_chemistry.algorithms import create
from qdk_chemistry.data import Orbitals, Structure

# Create H2 molecule
coords = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.4]])
structure = Structure(coords, ["H", "H"])

# Obtain orbitals from an SCF calculation
scf_solver = create("scf_solver")
E_scf, wfn = scf_solver.run(
    structure, charge=0, spin_multiplicity=1, basis_information="sto-3g"
)
orbitals = wfn.get_orbitals()

print(f"SCF Energy: {E_scf:.6f} Hartree")
# end-cell-create
################################################################################

################################################################################
# start-cell-access
# Access orbital coefficients (returns tuple of alpha/beta matrices)
coeffs_alpha, coeffs_beta = orbitals.get_coefficients()
print(f"Orbital coefficients shape: {coeffs_alpha.shape}")

# Access orbital energies (returns tuple of alpha/beta vectors)
energies_alpha, energies_beta = orbitals.get_energies()
print(f"Orbital energies: {energies_alpha}")

# Access atomic orbital overlap matrix
ao_overlap = orbitals.get_overlap_matrix()
print(f"AO overlap matrix shape: {ao_overlap.shape}")

# Access basis set information
basis_set = orbitals.get_basis_set()
print(f"Basis set: {basis_set.get_name()}")

# Check calculation type
is_restricted = orbitals.is_restricted()
print(f"Is restricted: {is_restricted}")

# Get size information
num_mos = orbitals.get_num_molecular_orbitals()
num_aos = orbitals.get_num_atomic_orbitals()
print(f"MOs: {num_mos}, AOs: {num_aos}")

# Get summary
print(orbitals.get_summary())
# end-cell-access
################################################################################

################################################################################
# start-cell-serialization
# Use a temporary directory for file I/O
with tempfile.TemporaryDirectory() as tmpdir:
    tmpdir_path = Path(tmpdir)
    orbitals_file = tmpdir_path / "molecule.orbitals.json"

    # Generic serialization with format specification
    orbitals.to_file(str(orbitals_file), "json")
    orbitals_from_file = Orbitals.from_file(str(orbitals_file), "json")

    # JSON serialization
    orbitals.to_json_file(str(orbitals_file))
    orbitals_from_json_file = Orbitals.from_json_file(str(orbitals_file))

    # Direct JSON conversion
    j = orbitals.to_json()
    orbitals_from_json = Orbitals.from_json(j)

    # HDF5 serialization (if available)
    # orbitals.to_hdf5_file(str(tmpdir_path / "molecule.orbitals.h5"))
    # orbitals_from_hdf5 = Orbitals.from_hdf5_file(str(tmpdir_path / "molecule.orbitals.h5"))
# end-cell-serialization
################################################################################
